### DISCUSSION TITLE
z.string().url() validates invalid urls

### DISCUSSION THREAD

**Discussion by slackerzz** [colinhacks/zod#3190] [Q&A] [answered]
zod: 3.22.4
nodejs: 20.11.0

I've encountered a couple of validations issue with urls:

```js
import { z } from 'zod';

const schema = z.string().url();

const res = schema.safeParse("https://example");

console.log('success', res.success)
console.log('error', res.error?.issues ?? [])
```

will return:
```
success true
error []
```

another case is:

```js
import { z } from 'zod';

const schema = z.string().url();

const res = schema.safeParse("https://example.com/a b");

console.log('success', res.success)
console.log('error', res.error?.issues ?? [])
```

that again will return:
```
success true
error []
```

**Comment by JacobWeisenburger** [comment]
These are both valid URLs:

```
https://example
https://example.com/a b
```

Why do you think they are not valid?

**Comment by slackerzz** [comment]
Hi, @JacobWeisenburger, thank you for your answer.
`https://example` doesn't have a top-level domain, but you are right. `http://localhost` for example is also valid. What I need here is a different validation.

`https://example.com/a b` needs to be URL encoded to be valid, if you paste it in the browser it will be automatically encoded as `https://example.com/a%20b`. But if you try with curl:
```
curl "https://example.com/a b"
```
you will get:
```
curl: (3) URL rejected: Malformed input to a URL function
```

However, I've seen that the validation is based on catching errors while calling `new URL(...)`, this function automatically takes care of the encoding.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const schema = z.string().url().transform( url => new URL( url ).href )
console.log( schema.parse( 'https://example.com/a b' ) ) // https://example.com/a%20b
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
